Migrate the vimeo shortcode AMP logic to Jetpack from the AMP plugin

kienstra opened the pull request:
#### Summary 
<!--- Provide a general summary of your changes in the Title above -->
Proposes migrating the AMP conversion of the Jetpack `[vimeo]` shortcode from the [AMP plugin](https://github.com/ampproject/amp-wp) to Jetpack.

Fixes https://github.com/ampproject/amp-wp/issues/3309

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
Would you be open to migrating the [AMP plugin's conversion](https://github.com/ampproject/amp-wp/blob/ea9e6fb9d262e699ea64978794840ba9868715f6/includes/embeds/class-amp-vimeo-embed.php#L71) of Jetpack's `[vimeo]` shortcode into Jetpack?

This should give you more control over how your shortcode works.

As Weston mentioned in https://github.com/ampproject/amp-wp/issues/3309#issuecomment-533255590, there's been [ongoing work](https://github.com/ampproject/amp-wp/issues/1021) to move Jetpack functionality from the AMP plugin into Jetpack.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
This is a new feature to Jetpack, but mainly a migration of the AMP plugin's logic into Jetpack. 

#### Testing instructions:
1. Ensure that `wp-config.php` has `define( 'JETPACK_DEV_DEBUG', true);`
2. In `/wp-admin`, in the Jetpack 'Settings' page, and in the 'Writing' tab, ensure this is toggled on:

<img width="818" alt="jetpack-here" src="https://user-images.githubusercontent.com/4063887/67992321-dded6400-fc01-11e9-81ea-9a05716d1eca.png">

3. Fetch this PR's branch
4. Clone the AMP plugin 
```
$ git clone --recursive https://github.com/ampproject/amp-wp.git amp && cd amp
```
5. Fetch this PR of the AMP plugin, which deactivates the AMP plugin's `[vimeo]` shortcode callback: https://github.com/ampproject/amp-wp/pull/3678
6. Do `$ npm install && composer install`
7. Activate the AMP plugin
8. Create a new post with a Shortcode block
9. Add a shortcode, like:
```
[vimeo id="84618094" width="600" height="400"]
```
10. Preview the front-end of the AMP URL, like by adding `&amp` or `?amp`to the URL, and look at how the Vimeo shortcode looks
11. `checkout` the `master` branch of Jetpack, not this PR's branch
12. Preview the front-end of the AMP URL, and notice how the Vimeo shortcode looks
13. Expected: The shortcode should look the same, whether using this PR's branch, or the Jetpack plugin's `master` branch
14. Here are some more shortcodes to test:
```
[vimeo id="84618094"]
[vimeo 8461809]
[vimeo id="84618094" width="600"]
[vimeo id="84618094" width="400" height="300"]
[vimeo https://vimeo.com/84618094]
```

## Before

<img width="917" alt="before-vimeo-shortcode" src="https://user-images.githubusercontent.com/4063887/67992159-50aa0f80-fc01-11e9-87b3-3ddcf408f27a.png">

## After

<img width="899" alt="after-vimeo-shortcode" src="https://user-images.githubusercontent.com/4063887/67992178-5bfd3b00-fc01-11e9-94b4-abf0751fe18a.png">

#### Proposed changelog entry for your changes:
Migrate AMP-conversion of Vimeo shortcode from the AMP plugin to Jetpack.

kienstra committed:
Migrate the vimeo shortcode AMP logic from AMP plugin, add tests

Most of the functionality is the same,
though there are some changes.
Also, add 2 parameters to the filter
'video_embed_html'.
This might be questionable,
but it prevents adding another filter,
or adding AMP logic directly in the
Vimeo shortcode file.

kienstra committed:
In response to Code Climate, shorten 2 functions

These functions were 2 and 1 lines long,
respectively.

kienstra committed:
Abstract part of filter_vimeo_shortcode() into another function

In response to Code Climate
report, with the cognitive complexity too high.

kienstra committed:
Move the shortcode logic into a new class

In response to a Code Climate
report of the class being too long.

kienstra committed:
Add new file to the files to be required

Also, add it to the
phpcs whitelist to be linted.

kienstra committed:
Update the PHP DocBlocks, correct typos

Instead of onthe, change this to on the.
Also, front-end instead of frontend.

kienstra commented:
With the commits above, this should be ready for review. But no pressure, this is a busy week 😄

chvillanuevap commented:
Does this work with just `[vimeo 84618094]`? Because that's supported by the current shortcode in Jetpack.

kienstra committed:
Handle a shortcode of [vimeo 12345]

This will then use 12345 as the data-videoid.
For compatibility with how the non-AMP
shortcode works,
though this didn't work in AMP before.

kienstra commented:
Hi @chvillanuevap, it does now 😄

kienstra committed:
Rename Jetpack_AMP_Shortcodes to Jetpack_AMP_Vimeo_Shortcode

In case there are more shortcode handlers.
The Code Climate limit of around 200 LOC
per class won't allow for more than 1 more
shortcode handler in this class.

kienstra committed:
In response to Code Climate, remove build_attributes_string()

This is only called by build_tag(),
so it shouldnt' be needed.

kienstra committed:
Reduce lines of code in response to Code Climate

There were apparently 5 too many LOC,
so this reduces the lines.

kienstra committed:
Rename $string to $collected_attributes

The name of $string is misleading,
as it's actually an array().

kienstra committed:
Empty commit to trigger another Travis build

kienstra committed:
Merge branch 'master' into add/vimeo-shortcode-amp

kienstra commented:
There's still a [failing Travis job](https://travis-ci.org/Automattic/jetpack/jobs/606147782?utm_medium=notification&utm_source=github_status). From the error message, I can't tell if it's caused by this PR, or how to fix it:

```
$ ./tests/run-travis.sh
Travis CI command: yarn test-dangerci-and-adminpage-and-extensions
yarn run v1.3.2
(node:6191) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
$ yarn danger ci && yarn test-adminpage && yarn test-extensions
(node:6213) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
$ /home/travis/build/Automattic/jetpack/node_modules/.bin/danger ci
Failing the build, there is 1 fail.
Feedback: https://github.com/Automattic/jetpack/pull/13921#issuecomment-548609003
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
The command "./tests/run-travis.sh" exited with 1.

```

kienstra committed:
In response to failed Travis build, change file in whitelist

This file name changed,
so change it in the whitelist also.

kienstra committed:
Rename filter_vimeo_shortcode() to filter_shortcode()

This is in a class with Vimeo in the name,
so it's probably not necessary to have
Vimeo in the function name.

kienstra committed:
Move test_init() above a dataProvider for a lower function

This should have been almost at the top
before, so move it one place up.

kienstra committed:
Replace [] syntax for arrays with array()

It looks like array() is by far
the most common syntax in Jetpack.

jeherve commented on 3rd-party/3rd-party.php:
```diff
@@ -13,6 +13,7 @@
 	'bitly.php',
 	'buddypress.php',
 	'class.jetpack-amp-support.php',
+	'class.jetpack-amp-vimeo-shortcode.php',
```
Instead of being its own file, I wonder if it wouldn't be best to have this part of the existing Vimeo shortcode handler in `modules/shortcodes/vimeo.php`.

This may also allow us to rely on the existing work done there to extract data from Vimeo URLs.

kienstra replied:
Good idea. The commits below apply it, including https://github.com/Automattic/jetpack/pull/13921/commits/c66aff4e3f24950ca57b5ab92598c6e2b66bf20e. But it looks like I still need to fix unit tests.

jeherve commented on 3rd-party/class.jetpack-amp-vimeo-shortcode.php:
```diff
@@ -0,0 +1,133 @@
+	 * @return string The filtered HTML.
+	 */
+	public static function filter_shortcode( $html, $shortcode_tag, $attr ) {
+		if ( ! Jetpack_AMP_Support::is_amp_request() || 'vimeo' !== $shortcode_tag ) {
```
Whenever we use the `Jetpack_AMP_Support` class, we try to check if it exists first, as it may not always exist in all environments (such as WordPress.com, where this code will be synchronized).

kienstra replied:
Good idea. It's applied in https://github.com/Automattic/jetpack/pull/13921/commits/8afd00617c8bc4fce986176f3f8cc56c120d14e7

jeherve commented on 3rd-party/class.jetpack-amp-vimeo-shortcode.php:
```diff
@@ -0,0 +1,133 @@
+		}
+
+		$aspect_ratio   = 0.5625;
+		$default_width  = 600;
```
Do you think you could add some comments clarifying the choices for those default values?

kienstra replied:
That would be nice 😄 

But I'm not sure about them, as they were [copied from the AMP plugin](https://github.com/ampproject/amp-wp/blob/ea9e6fb9d262e699ea64978794840ba9868715f6/includes/embeds/class-amp-vimeo-embed.php#L17). 

The `0.5625` aspect ratio is 9 x 16 (usually referred to as `16 x 9`), and that's a very common ratio for videos.

jeherve replied:
Interesting. I couldn't find more, not even in the PR that first introduced those dimensions.  ¯\_(ツ)_/¯
 https://github.com/ampproject/amp-wp/pull/702

jeherve commented on bin/phpcs-whitelist.js:
```diff
@@ -2,6 +2,7 @@
 module.exports = [
 	'3rd-party/3rd-party.php',
 	'3rd-party/class.jetpack-amp-support.php',
+	'3rd-party/class.jetpack-amp-vimeo-shortcode.php',
```
Thanks for adding this in!

kienstra replied:
Sure!

jeherve commented:
Thanks a lot for the contribution. I left a few comments above to get us started.

kienstra committed:
Check that the class exists before calling is_amp_request()

As Jeremy mentioned,
it's possible that the class won't exist.

kienstra committed:
Move logic from Jetpack_AMP_Vimeo_Shortcode into shortcodes/vimeo.php

As Jeremy mentioned,
this is probably better in that file.
It relates directly to that file,
which handles Vimeo shortcode.

kienstra committed:
Address 2 Code Climate issues

It looks like amp_vimeo_shortcode() was
too long, and had too much complexity.

kienstra committed:
Merge branch 'master' into add/vimeo-shortcode-amp

kienstra committed:
Code Climate: Attempt to reduce 'Cognitive Complexity'

This was rated 6, where
the highest allowed is 5.

kienstra committed:
Remove references to removed file

class.jetpack-amp-vimeo-shortcode.php was deleted,
so don't require it anymore,
or do phpcs checks for it.

kienstra committed:
Add parent::tearDown() call, correct @covers tags

Maybe this was responsible for some failing
unit tests.

kienstra committed:
Give $video_id an initial value of null

So it won't be passed to compact
as undefined.

kienstra committed:
Stop using build_tag(), and instead simply construct the HTML

Use a normal sprintf() call
instead of using that helper method.

kienstra committed:
Remove needless extra line, correct name of tests

The tests are now for amp_vimeo_shortcode(),
instead of the previous method.

jeherve commented on modules/shortcodes/vimeo.php:
```diff
@@ -172,6 +172,77 @@ function vimeo_shortcode( $atts ) {
+		$video_id = jetpack_shortcode_get_vimeo_id( $attr );
+	} elseif ( isset( $attr['id'] ) ) {
+		$video_id = $attr['id'];
+	}
```
Should we bail if we don't have a video ID by that point, or if it is null or `0`, instead of waiting until later?

kienstra replied:
Good idea, applied in https://github.com/Automattic/jetpack/pull/13921/commits/cb09ba40f08edf736c752d7bc6cc06be1a30e5e4#diff-35236fb9fce8df90f9eea45af69cc92cR199

jeherve commented on modules/shortcodes/vimeo.php:
```diff
@@ -172,6 +172,77 @@ function vimeo_shortcode( $atts ) {
+		$height = isset( $attr['height'] ) ? $attr['height'] : $default_height;
+	}
+
+	return array( $width, $height );
```
Do we want to run those values through `absint` here?

kienstra replied:
Good idea, https://github.com/Automattic/jetpack/pull/13921/commits/0a6732aa9d5c14f472e2033a5ea8e54c8a28a6a8 applies it

jeherve commented on modules/shortcodes/vimeo.php:
```diff
@@ -172,6 +172,77 @@ function vimeo_shortcode( $atts ) {
+		( class_exists( 'Jetpack_AMP_Support' ) && ! Jetpack_AMP_Support::is_amp_request() )
+		||
+		'vimeo' !== $shortcode_tag
+	) {
```
If `Jetpack_AMP_Support` does not exist at all, we'll end up continuing with the function. Ideally we want to cover 3 use-cases:

- A different shortcode tag.
- No `Jetpack_AMP_Support` class.
- Not an AMP request.

If any of those is true, we bail.

kienstra replied:
Sounds good, https://github.com/Automattic/jetpack/pull/13921/commits/4e3a04708514fa918bc75de6ab67b0882e19754c changes this so that it returns if any of those 3 cases is true.

jeherve commented on modules/shortcodes/vimeo.php:
```diff
@@ -172,6 +172,77 @@ function vimeo_shortcode( $atts ) {
+		return $html;
+	}
+
+	$video_id = null;
```
To stay consistent it may be best to default to `0` here, like we do in `jetpack_shortcode_get_vimeo_id`.

kienstra replied:
Sounds good, https://github.com/Automattic/jetpack/pull/13921/commits/ce216622af3f62de61ef55deb5cad801b9866509 changes this

jeherve commented on modules/shortcodes/vimeo.php:
```diff
@@ -172,6 +172,77 @@ function vimeo_shortcode( $atts ) {
+	$video_id = null;
+	if ( isset( $attr[0] ) ) {
+		$video_id = jetpack_shortcode_get_vimeo_id( $attr );
+	} elseif ( isset( $attr['id'] ) ) {
```
Should we check for `! empty()` instead to cover invalid values like `0`?

kienstra replied:
Sure, https://github.com/Automattic/jetpack/pull/13921/commits/ce216622af3f62de61ef55deb5cad801b9866509 checks for `! empty()`.

jeherve commented on modules/shortcodes/vimeo.php:
```diff
@@ -172,6 +172,77 @@ function vimeo_shortcode( $atts ) {
+		$video_id = $attr['id'];
+	}
+
+	list( $width, $height ) = jetpack_get_amp_vimeo_dimensions( $attr );
```
Instead of doing a `list` followed by a `compact`, I wonder if it wouldn't be clearer / cheaper to create an array in the beginning and merge / push each value to it.

kienstra replied:
This should be resolved through https://github.com/Automattic/jetpack/pull/13921/files/66e25acc900c5ec781765a06c15c640c0c61374b#diff-35236fb9fce8df90f9eea45af69cc92c, but let me know not.

jeherve commented on modules/shortcodes/vimeo.php:
```diff
@@ -172,6 +172,77 @@ function vimeo_shortcode( $atts ) {
+	}
+
+	list( $width, $height ) = jetpack_get_amp_vimeo_dimensions( $attr );
+	return jetpack_render_amp_vimeo( compact( 'video_id', 'width', 'height' ) );
```
I realize you may have done this because of a CodeClimate recommendation, but I'd be willing to ignore a CodeClimate warning about a too long function here if that could help us simplify things by getting rid of `jetpack_render_amp_vimeo` and returning the `amp-vimeo` tag here.

kienstra replied:
Thanks, that's a good idea. It's applied in https://github.com/Automattic/jetpack/pull/13921/commits/cb09ba40f08edf736c752d7bc6cc06be1a30e5e4

kienstra committed:
Remove jetpack_render_amp_vimeo(), move it into amp_vimeo_shortcode()

As Jeremy mentioned,
it's not needed to have this function.
it's only called in that function.

kienstra committed:
Make $video_id 0 instead of null

As Jeremy mentioned,
that keeps it consistent with
jetpack_shortcode_get_vimeo_id.

kienstra committed:
Also exit if Jetpack_AMP_Support doesn't exist

Per Jeremy's code review comment.

kienstra committed:
Use absint() for the width and height

Ensure these are int values,
before using them in amp_vimeo_shortcode().

kienstra committed:
Add a unit test for jetpack_get_amp_vimeo_dimensions()

To ensure this works as expected.

kienstra committed:
In sprintf(), change %d to %s

Because the values it will interpolate
are wrapped in esc_attr(), they
will be strings.

kienstra commented:
The commits above should address all of the code review points.

jeherve reviewed (changes requested):
I have a few more comments. All in all, I think something like this could work; I tried to gather all comments I left you. This would mean updating the tests though

<details>
  <summary>diff below</summary>

```diff

diff --git a/modules/shortcodes/vimeo.php b/modules/shortcodes/vimeo.php
index b7210ff68..20fd74f0d 100644
--- a/modules/shortcodes/vimeo.php
+++ b/modules/shortcodes/vimeo.php
@@ -37,74 +37,70 @@ function jetpack_shortcode_get_vimeo_id( $atts ) {
 }
 
 /**
- * Convert a Vimeo shortcode into an embed code.
+ * Get video dimensions.
  *
- * @param array $atts An array of shortcode attributes.
+ * @since 8.0.0
  *
- * @return string The embed code for the Vimeo video.
+ * @param array $attr     The attributes of the shortcode.
+ * @param array $old_attr Optional array of attributes from the old shortcode format.
+ *
+ * @return array Width and height.
  */
-function vimeo_shortcode( $atts ) {
+function jetpack_shortcode_get_vimeo_dimensions( $attr, $old_attr = array() ) {
 	global $content_width;
 
-	$attr = array_map(
-		'intval',
-		shortcode_atts(
-			array(
-				'id'       => 0,
-				'width'    => 0,
-				'height'   => 0,
-				'autoplay' => 0,
-				'loop'     => 0,
-			),
-			$atts
-		)
-	);
-
-	if ( isset( $atts[0] ) ) {
-		$attr['id'] = jetpack_shortcode_get_vimeo_id( $atts );
-	}
-
-	if ( ! $attr['id'] ) {
-		return '<!-- vimeo error: not a vimeo video -->';
-	}
-
-	// Handle old shortcode params such as h=500&w=350.
-	$params = shortcode_new_to_old_params( $atts );
-	$params = str_replace( array( '&amp;', '&#038;' ), '&', $params );
-	parse_str( $params, $args );
-
-	$width  = intval( $attr['width'] );
-	$height = intval( $attr['height'] );
+	$default_width  = 600;
+	$default_height = 338;
+	$aspect_ratio   = $default_height / $default_width;
+	$width          = ( ! empty( $attr['width'] ) ? absint( $attr['width'] ) : $default_width );
+	$height         = ( ! empty( $attr['height'] ) ? absint( $attr['height'] ) : $default_height );
 
-	// Support w and h argument as fallback.
-	if ( empty( $width ) && isset( $args['w'] ) ) {
-		$width = intval( $args['w'] );
+	/*
+	 * Support w and h argument as fallbacks.
+	 */
+	if (
+		$default_width === $width
+		&& ! empty( $old_attr['w'] )
+	) {
+		$width = absint( $old_attr['w'] );
 
-		if ( empty( $height ) && ! isset( $args['h'] ) ) {
-			// The case where w=300 is specified without h=200, otherwise $height
-			// will always equal the default of 300, no matter what w was set to.
-			$height = round( ( $width / 640 ) * 360 );
+		if (
+			$default_width === $width
+			&& empty( $old_attr['h'] )
+		) {
+			$height = round( $width * $aspect_ratio );
 		}
 	}
 
-	if ( empty( $height ) && isset( $args['h'] ) ) {
-		$height = (int) $args['h'];
+	if (
+		$default_height === $height
+		&& ! empty( $old_attr['h'] )
+	) {
+		$height = absint( $old_attr['h'] );
 
-		if ( ! isset( $args['w'] ) ) {
-			$width = round( ( $height / 360 ) * 640 );
+		if ( empty( $old_attr['w'] ) ) {
+			$width = round( $height * $aspect_ratio );
 		}
 	}
 
-	if ( ! $width && ! empty( $content_width ) ) {
+	/*
+	 * If we have a content width defined, let it be the new default.
+	 */
+	if (
+		$default_width === $width
+		&& ! empty( $content_width )
+	) {
 		$width = absint( $content_width );
 	}
 
-	// If setting the width with content_width has failed, defaulting.
-	if ( ! $width ) {
-		$width = 640;
-	}
-
-	if ( ! $height ) {
+	/*
+	 * If we have a custom width, we need a custom height as well
+	 * to maintain aspect ratio.
+	 */
+	if (
+		$default_width !== $width
+		&& $default_height === $height
+	) {
 		$height = round( ( $width / 640 ) * 360 );
 	}
 
@@ -130,6 +126,46 @@ function vimeo_shortcode( $atts ) {
 	 */
 	$height = (int) apply_filters( 'vimeo_height', $height );
 
+	return array( $width, $height );
+}
+
+/**
+ * Convert a Vimeo shortcode into an embed code.
+ *
+ * @param array $atts An array of shortcode attributes.
+ *
+ * @return string The embed code for the Vimeo video.
+ */
+function vimeo_shortcode( $atts ) {
+	$attr = array_map(
+		'intval',
+		shortcode_atts(
+			array(
+				'id'       => 0,
+				'width'    => 0,
+				'height'   => 0,
+				'autoplay' => 0,
+				'loop'     => 0,
+			),
+			$atts
+		)
+	);
+
+	if ( isset( $atts[0] ) ) {
+		$attr['id'] = jetpack_shortcode_get_vimeo_id( $atts );
+	}
+
+	if ( ! $attr['id'] ) {
+		return '<!-- vimeo error: not a vimeo video -->';
+	}
+
+	// Handle old shortcode params such as h=500&w=350.
+	$params = shortcode_new_to_old_params( $atts );
+	$params = str_replace( array( '&amp;', '&#038;' ), '&', $params );
+	parse_str( $params, $args );
+
+	list( $width, $height ) = jetpack_shortcode_get_vimeo_dimensions( $attr, $args );
+
 	$url = esc_url( 'https://player.vimeo.com/video/' . $attr['id'] );
 
 	// Handle autoplay and loop arguments.
@@ -169,65 +205,41 @@ function vimeo_shortcode( $atts ) {
 
 	return $html;
 }
-
 add_shortcode( 'vimeo', 'vimeo_shortcode' );
 
 /**
  * Filters the Vimeo shortcode to be AMP-compatible.
  *
- * @param string $html The shortcode HTML.
+ * @param string $html          The shortcode HTML.
  * @param string $shortcode_tag The shortcode's tag (name).
- * @param array  $attr The attributes of the shortcode.
+ * @param array  $attr          The attributes of the shortcode.
+ *
  * @return string The filtered HTML.
  */
-function amp_vimeo_shortcode( $html, $shortcode_tag, $attr ) {
-	if ( 'vimeo' !== $shortcode_tag || ! class_exists( 'Jetpack_AMP_Support' ) || ! Jetpack_AMP_Support::is_amp_request() ) {
+function jetpack_amp_vimeo_shortcode( $html, $shortcode_tag, $attr ) {
+	if (
+		'vimeo' !== $shortcode_tag
+		|| ! class_exists( 'Jetpack_AMP_Support' )
+		|| ! Jetpack_AMP_Support::is_amp_request()
+	) {
 		return $html;
 	}
 
-	$video_id = 0;
-	if ( isset( $attr[0] ) ) {
-		$video_id = jetpack_shortcode_get_vimeo_id( $attr );
-	} elseif ( ! empty( $attr['id'] ) ) {
-		$video_id = $attr['id'];
-	}
-
+	$video_id = ( ! empty( $attr['id'] ) ? $attr['id'] : jetpack_shortcode_get_vimeo_id( $attr ) );
 	if ( empty( $video_id ) ) {
-		return '';
+		return $html;
 	}
 
-	list( $width, $height ) = jetpack_get_amp_vimeo_dimensions( $attr );
+	list( $width, $height ) = jetpack_shortcode_get_vimeo_dimensions( $attr );
+
 	return sprintf(
-		'<amp-vimeo data-videoid="%s" layout="responsive" width="%s" height="%s"></amp-vimeo>',
+		'<amp-vimeo data-videoid="%s" layout="responsive" width="%d" height="%d"></amp-vimeo>',
 		esc_attr( $video_id ),
-		esc_attr( $width ),
-		esc_attr( $height )
+		absint( $width ),
+		absint( $height )
 	);
 }
-
-add_filter( 'do_shortcode_tag', 'amp_vimeo_shortcode', 10, 3 );
-
-/**
- * Gets the width and height of the AMP [vimeo] shortcode.
- *
- * @param array $attr The shortcode attributes.
- * @return array The dimensions, width at index 0, and height at index 1.
- */
-function jetpack_get_amp_vimeo_dimensions( $attr ) {
-	$default_width  = 600;
-	$default_height = 338;
-	$aspect_ratio   = $default_height / $default_width;
-
-	if ( ! empty( $GLOBALS['content_width'] ) ) {
-		$width  = $GLOBALS['content_width'];
-		$height = round( $GLOBALS['content_width'] * $aspect_ratio );
-	} else {
-		$width  = isset( $attr['width'] ) ? $attr['width'] : $default_width;
-		$height = isset( $attr['height'] ) ? $attr['height'] : $default_height;
-	}
-
-	return array( absint( $width ), absint( $height ) );
-}
+add_filter( 'do_shortcode_tag', 'jetpack_amp_vimeo_shortcode', 10, 3 );
 
 /**
  * Callback to modify output of embedded Vimeo video using Jetpack's shortcode.

```

</details>

kienstra committed:
Commit Jeremy's diff

Uses jetpack_shortcode_get_vimeo_dimensions()
for both the non-AMP and AMP shortcodes.

kienstra committed:
Update unit tests for the previous commit

The means of getting dimension changed,
so update the tests.

kienstra commented on modules/shortcodes/vimeo.php:
```diff
@@ -172,6 +172,63 @@ function vimeo_shortcode( $atts ) {
+
+	list( $width, $height ) = jetpack_get_amp_vimeo_dimensions( $attr );
+	return sprintf(
+		'<amp-vimeo data-videoid="%s" layout="responsive" width="%s" height="%s"></amp-vimeo>',
```
Sure, applied with your diff [here](https://github.com/Automattic/jetpack/pull/13921/commits/d874efe0e1d33de850f3d2e760b8e6484cf5c4a2#diff-35236fb9fce8df90f9eea45af69cc92cR238).

kienstra commented on modules/shortcodes/vimeo.php:
```diff
@@ -172,6 +172,63 @@ function vimeo_shortcode( $atts ) {
+	}
+
+	if ( empty( $video_id ) ) {
+		return '';
```
Sure, applied [here](https://github.com/Automattic/jetpack/pull/13921/commits/d874efe0e1d33de850f3d2e760b8e6484cf5c4a2#diff-35236fb9fce8df90f9eea45af69cc92cR230) with your diff.

kienstra commented on modules/shortcodes/vimeo.php:
```diff
@@ -172,6 +172,63 @@ function vimeo_shortcode( $atts ) {
+	}
+
+	$video_id = 0;
+	if ( isset( $attr[0] ) ) {
```
Sure, your diff applies that [here](https://github.com/Automattic/jetpack/pull/13921/commits/d874efe0e1d33de850f3d2e760b8e6484cf5c4a2#diff-35236fb9fce8df90f9eea45af69cc92cR228).

kienstra commented on modules/shortcodes/vimeo.php:
```diff
@@ -172,6 +172,63 @@ function vimeo_shortcode( $atts ) {
+	$default_height = 338;
+	$aspect_ratio   = $default_height / $default_width;
+
+	if ( ! empty( $GLOBALS['content_width'] ) ) {
```
Sure, addressed in your diff [here](https://github.com/Automattic/jetpack/pull/13921/commits/d874efe0e1d33de850f3d2e760b8e6484cf5c4a2#diff-35236fb9fce8df90f9eea45af69cc92cL221).

kienstra commented on modules/shortcodes/vimeo.php:
```diff
@@ -172,6 +172,63 @@ function vimeo_shortcode( $atts ) {
+ * @param array $attr The shortcode attributes.
+ * @return array The dimensions, width at index 0, and height at index 1.
+ */
+function jetpack_get_amp_vimeo_dimensions( $attr ) {
```
Sure, that's applied [here](https://github.com/Automattic/jetpack/pull/13921/commits/d874efe0e1d33de850f3d2e760b8e6484cf5c4a2#diff-35236fb9fce8df90f9eea45af69cc92cR49) in your diff.

kienstra commented on modules/shortcodes/vimeo.php:
```diff
@@ -172,6 +172,63 @@ function vimeo_shortcode( $atts ) {
+ * @param array  $attr The attributes of the shortcode.
+ * @return string The filtered HTML.
+ */
+function amp_vimeo_shortcode( $html, $shortcode_tag, $attr ) {
```
Sure, your diff applies that [here](https://github.com/Automattic/jetpack/pull/13921/commits/d874efe0e1d33de850f3d2e760b8e6484cf5c4a2#diff-35236fb9fce8df90f9eea45af69cc92cR219).

kienstra committed:
Update the name of the function in the test files

There were still a few places where the new name
of jetpack_amp_vimeo_shortcode() did not appear.

kienstra committed:
Remove @covers tag for jetpack_shortcode_get_vimeo_dimensions()

That has its own tests,
so it's not very helpful
to mark that as @covered.

kienstra committed:
Empty commit to trigger Travis build

There was only 1 failure:
rbenv: version 7.2 not installed

kienstra commented:
Hi Jeremy,
Thanks for your [diff above](https://github.com/Automattic/jetpack/pull/13921#pullrequestreview-312303851), I committed it verbatim. Also, I made a few changes, like updating tests.

All of the testing shortcodes in step 13 [here](https://github.com/Automattic/jetpack/pull/13921#issue-335207983) look good with AMP. 

They don't all look exactly the same as the non-AMP shortcodes, as the `<amp-vimeo>` has a `layout="responsive"`, meaning that it will expand to fill its container. 

That was also the case in the AMP plugin.

jeherve committed:
Move AMP rendering into main shortcode function

jeherve reviewed (dismissed):
This looks good to me. It should be good to merge.

After thinking about this some more, I committed 1bfc3e698fc386bd273c7033a3468f38847daac7 to simplify things a bit more. Now that we moved dimensions into their own function, I think that makes things a bit simpler and easier to follow. It should also be a bit better performance-wise.

jeherve committed:
Update tests to match 1bfc3e698fc386bd273c7033a3468f38847daac7

jeherve committed:
Skip some tests on WordPress.com.

The AMP plugin is not up to date on WordPress.com yet, we can't test everything there yet.
